peer SAS duplicate grant purging reference model

jni2000 opened the pull request:
Preliminary implementation of peer SAS duplicate grant purging reference model for review.

kate-harrison commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,95 @@
+        cbsds_to_remove.append(cbsd)
+
+  for cbsd in cbsds_to_remove:
+    cbsd_list.remove(cbsd)
```
We need to return new FAD objects which have the corresponding CBSDs removed, not just ignore the result...

jni2000 replied:
The   input FAD objects are modified and the corresponding CBSDs are remvoed from   the input FAD objects. This is to allow easy chaining of different pre-iap   models.

TaranFurnell commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,95 @@
+
+  # Creating list of cbsds from SAS UUT FAD object and SAS Test Harness FAD object
+  cbsd_list = []
+  for cbsds in sas_uut_fad.getCbsdRecords():
```
Since all that matters is CBSD reference ID (see Kate's comment below) you can just keep track of the number of occurrences of a given ID (in a dictionary) and only include CBSDs with a single occurrence in the output.

jni2000 replied:
Thanks.   The logic has been updated. The input parameters are modified and no output   is provided.

jni2000 commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,95 @@
+           # Check if fccId and cbsdSerialNumber match
+        if ((cbsd_a['registrationRequest']['fccId'] ==
+             cbsd_b['registrationRequest']['fccId']) and
+             (cbsd_a['registrationRequest']['cbsdSerialNumber'] ==
```
CBSD   reference ID is used now.

jni2000 commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,95 @@
+          # Check for overlapping grantrequests
+          for grant_in_a in cbsd_a['grantRequests']:
+            for grant_in_b in cbsd_b['grantRequests']:
+              if checkForOverlappingGrants(grant_in_a,grant_in_b):
```
The   logic has been updated to remove the CBSD if it is present in FAD records of   multiple SASs.

jni2000 commented:
Kate and Taran

We have addressed your review comments please continue. Please note: we will change to return new FADs as output explicitly for all purging reference models in the follow up drops.

jni2000 commented:
We have made the change to return the FADs after removing all purged grants. Please continue to review.

jni2000 commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,81 @@
+"""
+import copy
+from collections import Counter
+def interSasDuplicateGrantReferenceModel(input_sas_uut_fad, input_sas_test_harness_fads):
```
Will interSasDuplicateGrantPurgingReferenceModel better ? Or can you please suggest one ?

TaranFurnell replied:
interSasDuplicateGrantPurgeReferenceModel?

jni2000 replied:
ok, that was also my first catch ! :)

jni2000 replied:
Chanegd   the name to "interSasDuplicateGrantPurgeReferenceModel" as   suggested.

jni2000 commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,81 @@
+  Returns:
+    sas_uut_fad: A FullActivityDump object containing the FAD record of SAS UUT after purged grants removed.    
+    sas_test_harness_fads: A list of FullActivityDump object containing the FAD record of 
+      SAS test harness after purged grants removed.    
```
Also, please see the reposes in #475. In brief, the FAD style of outputs was decided by the working group in an early f2f design meeting as well as the subsequent weekly status meetings.

jni2000 replied:
Changes   done as per the approach of modifying the input parameters.

jni2000 commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,81 @@
+
+  Checks if a CBSD is registered with more than one SAS and removes the grants of the
+  CBSD if the grant frequencies overlap either fully or partially. If all grants are
+  removed from a CBSD then the CBSD is removed from the FAD Object.
```
Updated   the paragraph based on the recent updates.

jni2000 commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,81 @@
+  sas_uut_fad = copy.deepcopy(input_sas_uut_fad)
+  sas_test_harness_fads = copy.deepcopy(input_sas_test_harness_fads)
+
+  all_cbsd_ids= []
```
Thanks   for the code. Changes have been done accordingly and tested using the FAD   object in PR485.

jni2000 commented:
Taran

We have addressed all your latest review comments including using and modifying the input FADs directly for all pre-IAP purging modules. Please take a look.

Thanks

TaranFurnell commented on src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py:
```diff
@@ -0,0 +1,59 @@
+from collections import defaultdict
+
+def interSasDuplicateGrantPurgeReferenceModel(sas_uut_fad, sas_test_harness_fads):
+  """ Removes duplicate grants from FAD objects of SAS UUT and SAS test harnesses.
```
Change to "Removes CBSDs with grants from more than one SAS from FAD objects of SAS UUT and SAS test harnesses."

jni2000 replied:
Changed as suggested. Thanks.

jni2000 commented:
Thanks, will do.

On Mar 23, 2018 6:27 PM, Taran Furnell <notifications@github.com> wrote:

@TaranFurnell approved this pull request.

Please make one minor update to the doc-string. Otherwise approved.

________________________________

In src/harness/reference_models/inter_sas_duplicate_grant/inter_sas_duplicate_grant.py<https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/472#discussion_r176877861>:

> +#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+"""
+==================================================================================
+  This is a subset of the pre-IAP reference model which implements inter-SAS
+  duplicate CBSD removal. If a CBSD has registered with multiple SASs then the
+  CBSD is removed from the FAD objects of the respective SASs. The main function
+  is interSasDuplicateGrantPurgeReferenceModel().
+==================================================================================
+"""
+
+from collections import defaultdict
+
+def interSasDuplicateGrantPurgeReferenceModel(sas_uut_fad, sas_test_harness_fads):
+  """ Removes duplicate grants from FAD objects of SAS UUT and SAS test harnesses.


Change to "Removes CBSDs with grants from more than one SAS from FAD objects of SAS UUT and SAS test harnesses."

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/472#pullrequestreview-106664961>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKpZrMU5BOF5bErBkLZUrUVj17q6T6Qwks5thXa7gaJpZM4SpjzM>.
